src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -6428,15 +6428,36 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         }
 
         if (!s) {
-            std::set<std::string> not_cpython_builtin = {
-                "sin", "cos", "gamma", "tan", "asin", "acos", "atan"
+            std::set<std::string> module_math = {
+                "sin", "cos", "gamma", "tan", "asin", "acos", "atan", "exp"
             };
-            if (ASRUtils::IntrinsicFunctionRegistry::is_intrinsic_function(call_name)
-             && not_cpython_builtin.find(call_name) == not_cpython_builtin.end()) {
+            std::set<std::string> module_numpy = {
+                "exp", "exp2"
+            };
+            // Complex conditions because of functions which are in both modules
+            bool is_math_imported = current_scope->resolve_symbol("math") != nullptr;
+            bool is_numpy_imported = current_scope->resolve_symbol("numpy") != nullptr;
+            bool is_math_module = module_math.find(call_name)!= module_math.end();
+            bool is_numpy_module = module_numpy.find(call_name)!= module_numpy.end();
+            if (( is_math_module && !is_numpy_module && !is_math_imported ) ||
+                ( is_numpy_module && !is_math_module && !is_numpy_imported ) ||
+                ( is_math_module && is_numpy_module && !is_math_imported && !is_numpy_imported )) {
```

Comment:
The following seems to work, can you please check what's going on here?
```py
from math import exp
def main0():
    x: f64 =  1.0
    print(sin(x))

main0()
```

Reply:
Actually error handling with imported modules is something I'm having trouble with. I could not exactly understand the code related to it so came up with this. Can you help me out with this and https://github.com/lcompilers/lpython/pull/1712#discussion_r1175426581 ? How should I approach this?
What we have to do is raise an error if ```from math import exp``` or ```from numpy import exp2``` is not present, for both functions that are unique to these module (like ```exp2``` in numpy) and ones that are common to them. Also functions imported from ```math``` should only accept scalars, but I hope that I implemented that error correctly.

Reply:
```py
from math import sin
```
The above generates the following ASR:
<details>
<summary>ASR</summary>

```Clojure
(TranslationUnit
    (SymbolTable
        1
        {
            _global_symbols:
                (Module
                    (SymbolTable
                        116
                        {
                            sin:
                                (ExternalSymbol
                                    116
                                    sin
                                    3 sin
                                    math
                                    []
                                    sin
                                    Public
                                )
                        })
                    _global_symbols
                    [math]
                    .false.
                    .false.
                ),
            main_program:
                (Program
                    (SymbolTable
                        115
                        {

                        })
                    main_program
                    []
                    []
                ),
            math:
                (Module
                    ...
                )
        })
    []
)
```
</details>

As you can see the ExternalSymbol is created for the `Sin` in the scope, if it is not present then you can throw an error.
Also, you can see the `module_name` (math) in the ExternalSymbol. you can use it to choose the modules like `math` or `numpy`.

Reply:
You can use the same approach for functions accepts scalar or vectors as arguments.

Any issues, let me know.